### PR TITLE
feature/nn6

### PR DESCRIPTION
**Pull request by gbaydin**
New PR to keep track of ongoing ml work.

### CODE REVIEW

**Comment by gbaydin** [comment]
This PR introduced recurrent neural networks (a vanilla RNN and an LSTM), a TextDataset for tokenizing and converting text to tensors, and a language model example script that is working well. Merging now. Next step would be to implement transformers and other things.

**Comment by dsyme** [comment]
Fantastic

File: src/DiffSharp.Core/Model.fs [hunk 1]
```diff
@@ -792,4 +808,200 @@ type VAE(xDim:int, zDim:int, ?hDims:seq<int>, ?nonlinearity:Tensor->Tensor, ?non
...
+            let v = value.[i]
+            r.hidden <- dsharp.matmul(v, wih.value) + dsharp.matmul(h.value, whh.value)
+            if bias then r.hidden <- r.hidden + b.value
+            r.hidden <- nonlinearity r.hidden
+            output.[i] <- r.hidden
+        let output = dsharp.stack output
+        if batchFirst then output.transpose(0, 1) else output
+
+/// <summary>Unit cell of a long short-term memory (LSTM) recurrent neural network. Prefer using the RNN class instead, which can combine RNNCells in multiple layers.</summary>
+type LSTMCell(inFeatures, outFeatures, ?bias, ?batchFirst) =
```
Comment (dsyme) [inline src/DiffSharp.Core/Model.fs:850]: I recommend we should move each actual model out into a separate file.  This file is getting too long

Comment (gbaydin) [reply src/DiffSharp.Core/Model.fs:850]: Agreed

**Pull request merged**